[flang][mlir][OpenMP] Add support for COPYPRIVATE

luporl opened the pull request:
Add initial handling of OpenMP COPYPRIVATE clause in Flang.

MLIR's omp.single operation was modified to support an optional
CopyPrivateVarList. It consists of pairs of variables and
functions. When present, each thread variable is updated with the
variable value of the thread that executed the single region,
using the specified functions to perform the copy.

When lowering COPYPRIVATE, Flang then generates the copy function
needed by each variable and builds the appropriate
CopyPrivateVarList. The translation to LLVM IR is done in
OMPIRBuilder, by calling createCopyPrivate() for each variable in
the list, which generates calls to __kmpc_copyprivate.

Fixes https://github.com/llvm/llvm-project/issues/63933

luporl commented:
Some refactoring was needed in `Bridge.cpp`, especially in `createHostAssociateVarClone()`
and `copyHostAssociateVar()`. It was needed to be able to clone and copy variables that may
not be host associated. But I guess the refactoring was not done in the best way, so any
suggestion of how to improve it is welcome.

Also, this implementation considers only single operations inside (or not) a parallel region.
If there are other OpenMP directives that provide parallelism without using parallel, where
single can be used, these will need to be handled later.

luporl commented on flang/lib/Lower/Bridge.cpp:
```diff
@@ -609,125 +610,41 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+  Fortran::lower::SymbolBox
+  createVarClone(const Fortran::semantics::Symbol &sym) override final {
+    fir::ExtendedValue exv = cloneSymbolValue(sym);
+    Fortran::lower::SymMap symMap;
+    addSymbol(sym, exv, /*forced=*/true, symMap);
+    return symMap.shallowLookupSymbol(sym);
```
I couldn't find a better way of converting an ExtendedValue into a SymbolBox.
The SymbolBox is needed later, in `copyVar()`, and is easily convertible into an ExtendedValue, when needed.

luporl commented on flang/test/Lower/OpenMP/copyprivate.f90:
```diff
@@ -0,0 +1,244 @@
+!CHECK:           hlfir.assign %[[K]] to %[[K_SYNC]]#0 temporary_lhs : i32, !fir.ref<i32>
+!CHECK-NEXT:      omp.terminator
+!CHECK-NEXT:    }
+!CHECK-NEXT:    %[[K:.*]] = fir.load %[[K_SYNC]]#0 : !fir.ref<i32>
+!CHECK-NEXT:    omp.barrier
+!CHECK-NEXT:    hlfir.assign %[[K]] to %{{.*}}#1 temporary_lhs : i32, !fir.ref<i32>
+!CHECK-NEXT:    omp.barrier
```
In my tests, the private variables were synchronized correctly only with a barrier after the load and another one after the assignment. It should be possible to group the loads and the assignments, however, to use only 2 barriers when there are multiple items in a copyprivate clause.

kiranchandramohan replied:
Is this now two per entry in the copyprivate clause?

luporl replied:
Yes. I'll try to group the loads and assigns, to reduce it to two per copyprivate clause.
Actually, the `omp.barrier` could probably be replaced by a regular memory barrier here, but I couldn't find a `fir` operation for it.

In fact, I'm a bit surprised these barriers are required at all. I expected the end of `single` to synchronize memory among all threads and that no more barriers would be needed after it, for the loads and stores in each thread.
But in practice, the lack of a barrier after an allocatable array load makes its old extent be used in `hlfir.assign`, resulting in a runtime error, and the lack of a barrier after the assign causes old values to be used in a subsequent print statement, for instance.
I have only noticed this issue with allocatable arrays so far.

luporl replied:
Hmm, maybe an `omp.flush` will do it.

NimishMishra replied:
@luporl Thanks for the patch. I went through the code and functionality and it looks fine to me, except for the fact that there are too many barriers. Correct me if I am wrong, but the following snippet seems to be the cause of 2 back-to-back barriers after load and assign:

```
        hlfir::Entity r2 = hlfir::loadTrivialScalar(loc, *builder, r);
        if (needBarrier && r2 != r)
          builder->create<mlir::omp::BarrierOp>(loc);
        builder->create<hlfir::AssignOp>(
            loc, r2, l,
            /*isWholeAllocatableAssignment=*/false,
            /*keepLhsLengthInAllocatableAssignment=*/false,
            /*temporary_lhs=*/true);
        if (needBarrier)
          builder->create<mlir::omp::BarrierOp>(loc);
```

However, I am not sure whether reducing this to one barrier may still cause one barrier per copyprivate list operation (since `copyVar` is called on each symbol). As such, does it make sense to factor out generation of barrier too from `copyVar` and explicitly generate a barrier at the end? Or maybe `needBarrier=false` for all-but-last variable in the copyprivate list? 

Rest all LGTM!

luporl replied:
@NimishMishra Thanks for the review! That's right, this is the part that inserts 2 barriers.
The problem is that, if I reduce it to one barrier, I get incorrect results when using copyprivate with allocatable arrays (tested on AArch64). It should be possible to refactor the code to generate all loads of the copyprivate list, then a barrier, then all assigns and then another barrier.

But, to be honest, I still don't understand why it only works with 2 barriers. I will try to investigate this further, to see if there is a better solution.

luporl replied:
I found out the issue with allocatable arrays was actually that the code that performs the deallocation of the sync var was ending up in the parallel region. The barriers were just preventing the sync var of being deallocated while still being used.

luporl replied:
The new implementation is going to use the `__kmpc_copyprivate` runtime function, which avoids the need to manually insert barriers.

NimishMishra commented on flang/lib/Semantics/resolve-directives.cpp:
```diff
@@ -2368,7 +2368,8 @@ void OmpAttributeVisitor::CheckDataCopyingClause(
           !(HasSymbolInEnclosingScope(symbol, currScope()) &&
-              symbol.test(Symbol::Flag::OmpPrivate))) {
+              (symbol.test(Symbol::Flag::OmpPrivate) ||
+                  symbol.test(Symbol::Flag::OmpFirstPrivate)))) {
```
Two points here:
1. Does this require a change to the message here? It says PRIVATE or THREADPRIVATE. 
2. Can we add a test for this change in semantic check?

luporl replied:
1. As FIRSTPRIVATE provides a superset of the functionality provided by PRIVATE, I thought that using it would already imply that the variable is private. I also noticed that gfortran allows FIRSTPRIVATE but doesn't mention it explicitly:
> copyprivate variable 'aaa' is not threadprivate or private in outer context

 Do you think it would be better to mention FIRSTPRIVATE explicitly in the error message?

2. Sure, I'll add a semantic check for it.

NimishMishra replied:
Yeah you are right: we don't need to explicitly call our firstprivate here.

kiranchandramohan commented:
@luporl I see that clang uses a runtime call to implement threadprivate. And that runtime call internally has a barrier.
https://openmp.llvm.org/doxygen/group__THREADPRIVATE.html#ga1453eca6136fd77e5de88ea0e78cc7a4

Have you explored using the runtime call?

In general, there is a desire to capture the privatisation and data-copying closes in the OpenMP dialect and then lower them later when we interface with the OpenMP IRBuilder. We did not do this initially to minimize any changes from the default FIR flow, lack of experience, and also the fact that privatisation in Parallel construct did not need a runtime call or any special construction. This will change with tasks (which need all the firstprivate variables to be allocated in the task data-structure). But copyprivate also could be a good candidate to initiate capturing this information in the OpenMP dialect and then lowering it later using the OpenMPIRBuilder since it can make use of the `kmpc_copyprivate` runtime call and also because it is only applicable to one construct.

Let me know what your thoughts are on this.

kiranchandramohan commented:
Also support for copyprivate was added to the OpenMPIRBuilder in the following patch.
https://reviews.llvm.org/D85617

jeanPerier reviewed (commented):
I am a bit rusty with OpenMP, but how do you ensure that the sync variable is shared by all the threads when this is a single in a subprogram without a parallel?

I would expect the alloca for the buffer to be "private" to each threads that entered the subprogram from a call in some parallel region (so unusable to broadcast the value to all threads).

e.g: do all threads print `1` in the program below?

```
subroutine foo(i)
 !$omp single
 i = 1
 !$omp end single copyprivate(i)
end subroutine

  integer, save :: i = 0
  !$omp threadprivate(i)
  !$omp parallel
  call foo(i)
  !$omp critical
  print *, i
  !$omp end critical 
  !$omp end parallel
end
```

From a design point of view, would there be advantages to keep the copyprivate implementation abstract in the IR (e.g, to not make the buffer/barriers/copies explicit yet)?

clementval commented:
> From a design point of view, would there be advantages to keep the copyprivate implementation abstract in the IR (e.g, to not make the buffer/barriers/copies explicit yet)?

+1
I think introducing abstraction in the OpenMP dialect and then have passes to lower that further down would be nice.

kiranchandramohan commented:
> I am a bit rusty with OpenMP, but how do you ensure that the sync variable is shared by all the threads when this is a single in a subprogram without a parallel?
> 
> I would expect the alloca for the buffer to be "private" to each threads that entered the subprogram from a call in some parallel region (so unusable to broadcast the value to all threads).

I think @jeanPerier is right here, storing the sync variable on the stack is probably not sufficient. I guess a global or support from runtime is required. This is probably the reason that the `kmpc_copyprivate` runtime exists.
https://github.com/llvm/llvm-project/blob/f42eb15c39405a8ae26d7f4f2fe7b29337ed6799/openmp/runtime/src/kmp_csupport.cpp#L2180

> would there be advantages to keep the copyprivate implementation abstract in the IR (e.g, to not make the buffer/barriers/copies explicit yet)?

One advantage is that it will help make it possible to handle copyprivate with a call to the runtime.

I think two options for this are :
1. Since the copyprivate clause is only applicable to the single construct we could consider making the `omp.single` operation an operation with two regions. The semantics being that the first region is executed only by the `single` thread and all other threads execute the `copy` region after the `single` region is executed. The copy region can later become the copy function used in the `kmpc_copyprivate` runtime call.
```
omp.single copyprivate(x1,x2) {
 x1 = bval1
 x2 = bval2
} copy  {
bb0(tmp1,tmp2): !tmp1 and tmp2 are block arguments corresponding to the sync variables of x1 and x2
   x1 = tmp1
   x2 = tmp2
}
```
2. Model it like the OpenMP reduction/OpenACC reduction + privatisation with a declare/recipe function that can later be used to generate the copy function that is needed in the kmpc_copyprivate runtime call.
```
omp.single copyprivate(x1:copy_x1,x2:copy_x2) {
 x1 = bval1
 x2 = bval2
}
```

luporl commented:
> I am a bit rusty with OpenMP, but how do you ensure that the sync variable is shared by all the threads when this is a single in a subprogram without a parallel?
> 
> I would expect the alloca for the buffer to be "private" to each threads that entered the subprogram from a call in some parallel region (so unusable to broadcast the value to all threads).
> 
> e.g: do all threads print `1` in the program below?
> 
> ```
> subroutine foo(i)
>  !$omp single
>  i = 1
>  !$omp end single copyprivate(i)
> end subroutine
> 
>   integer, save :: i = 0
>   !$omp threadprivate(i)
>   !$omp parallel
>   call foo(i)
>   !$omp critical
>   print *, i
>   !$omp end critical 
>   !$omp end parallel
> end
> ```
> 
This program doesn't compile:
```
test.f90:4:31: error: COPYPRIVATE variable 'i' is not PRIVATE or THREADPRIVATE in outer context
   !$omp end single copyprivate(i)
```

Making `i` threadprivate results in:
```
test.f90:2:22: error: A variable that appears in a THREADPRIVATE directive must be declared in the scope of a module or have the SAVE attribute, either explicitly or implicitly
   !$omp threadprivate(i)
```

And it is not possible to add the SAVE attribute to a dummy argument. But the following program shows the issue you are mentioning:
```
program main
  implicit none
  integer, save :: i = 0
  !$omp threadprivate(i)

  !$omp parallel
  call foo
  !$omp critical
  print *, i
  !$omp end critical
  !$omp end parallel

contains
  subroutine foo()
    !$omp single
    i = 1
    !$omp end single copyprivate(i)
  end subroutine
end program
```

Only one thread prints 1, all the others print 0. In this case the sync variable will indeed end up in a "private" thread stack, instead of the "shared" stack. So with this approach a global sync variable would be needed instead. Thanks for catching this.

luporl commented:
> There already a runtime Assign function but I think this operates on descriptors (
> 
> https://github.com/llvm/llvm-project/blob/fea023b129190edeb503dbe947034f925bbda666/flang/runtime/assign.cpp#L253
> ). So I guess the disadvantage here might be that we will always need descriptors.
> 
Ok, then I guess it's better to worry about generating multiple copy functions later, if this really becomes an issue.
I'm thinking about trying to implement option 2 and see if everything works properly. What do you think?

kiranchandramohan commented:
> > There already a runtime Assign function but I think this operates on descriptors (
> > https://github.com/llvm/llvm-project/blob/fea023b129190edeb503dbe947034f925bbda666/flang/runtime/assign.cpp#L253
> > 
> > ). So I guess the disadvantage here might be that we will always need descriptors.
> 
> Ok, then I guess it's better to worry about generating multiple copy functions later, if this really becomes an issue. I'm thinking about trying to implement option 2 and see if everything works properly. What do you think?

That would be great. Thanks in advance for trying this.

luporl committed:
[flang][mlir][OpenMP] Add support for COPYPRIVATE

Add initial handling of OpenMP COPYPRIVATE clause in Flang.

MLIR's omp.single operation was modified to support an optional
CopyPrivateVarList. It consists of pairs of variables and
functions. When present, each thread variable is updated with the
variable value of the thread that executed the single region,
using the specified functions to perform the copy.

When lowering COPYPRIVATE, Flang then generates the copy function
needed by each variable and builds the appropriate
CopyPrivateVarList. The translation to LLVM IR is done in
OMPIRBuilder, by calling createCopyPrivate() for each variable in
the list, which generates calls to __kmpc_copyprivate.

Fixes https://github.com/llvm/llvm-project/issues/63933

luporl commented on flang/lib/Lower/Bridge.cpp:
```diff
@@ -1075,16 +949,226 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+    return lookupSymbol(sym);
+  }
+
+  fir::ExtendedValue cloneSymbolValue(const Fortran::semantics::Symbol &sym) {
```
The new implementation approach (to be pushed soon) doesn't need to clone variables anymore.

luporl commented on flang/lib/Lower/Bridge.cpp:
```diff
@@ -1075,16 +949,226 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+
+  Fortran::lower::SymbolBox
+  getOriginalSymbolBox(const Fortran::semantics::Symbol &sym) {
+    const auto *details = sym.detailsIf<Fortran::semantics::HostAssocDetails>();
+    if (details) {
```
Removed in the new approach.

luporl commented on flang/lib/Lower/Bridge.cpp:
```diff
@@ -1075,16 +949,226 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+          auto if_builder = builder->genIfThenElse(loc, isAllocated);
+          if_builder.genThen([&]() {
+            std::string name = mangleName(sym) + ".alloc";
+            if (auto seqTy = symType.dyn_cast<fir::SequenceType>()) {
```
In the new approach this part of the code isn't touched anymore.

clementval commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -880,6 +884,156 @@ createReductionDecl(fir::FirOpBuilder &builder, llvm::StringRef reductionOpName,
+  bool inBox = false;
+};
+
+std::string TypeInfo::typeScan(mlir::Type ty) {
```
Can you use `getTypeAsString` from `FIRType.h`? It looks pretty similar. 

https://github.com/llvm/llvm-project/blob/2c257cf8721a030af03bd84f72e77864a7cf561b/flang/lib/Optimizer/Dialect/FIRType.cpp#L514

luporl replied:
Yes, it works. I wish I had found this earlier.

luporl commented:
A few notes about this last commit:

- I have implemented COPYPRIVATE support for HLFIR only. Do we also need to support it in FIR mode?

- I have written only a basic test for now, covering simple scalars and arrays. My plan is to add more tests, but I wanted to get some feedback first, to avoid having to rewrite the tests too much.

- There is one thing that I've noticed a bit late. The `__kmpc_copyprivate` function uses 2 barriers, which may impact the performance of COPYPRIVATE with multiple variables. Combining the updates of all variables to use a single `__kmpc_copyprivate` call is not straightforward.

clementval commented:
> * I have implemented COPYPRIVATE support for HLFIR only. Do we also need to support it in FIR mode?

I think HLFIR only is fine. We also only consider HLFIR from now in the OpenACC lowering.

kiranchandramohan commented:
> > * I have implemented COPYPRIVATE support for HLFIR only. Do we also need to support it in FIR mode?
> 
> I think HLFIR only is fine. We also only consider HLFIR from now in the OpenACC lowering.

+1. HLFIR flow only is fine, this is the default.

kiranchandramohan commented:
> * I have written only a basic test for now, covering simple scalars and arrays. My plan is to add more tests, but I wanted to get some feedback first, to avoid having to rewrite the tests too much.

I have not had a chance to go through in detail, I will do so later today or tomorrow. But overall this is a great improvement and in the direction that we would like to move forward. So thanks a lot for your effort here. If you can add a Fortran -> llvm ir test in flang/test/Integration with 2 copyprivates that would be great for review.

kiranchandramohan commented:
@psoni2628 This patch will be a very useful reference when you implement the firstprivatisation of tasks in the task datastructure.

luporl committed:
Fix mlir/test/Dialect/OpenMP/invalid.mlir

clementval commented:
> But overall this is a great improvement and in the direction that we would like to move forward.

+1. Great to see this direction being taken!

kiranchandramohan reviewed (commented):
> * There is one thing that I've noticed a bit late. The `__kmpc_copyprivate` function uses 2 barriers, which may impact the performance of COPYPRIVATE with multiple variables. Combining the updates of all variables to use a single `__kmpc_copyprivate` call is not straightforward.

Looking at the code, if the OpenMPIRBuilder provides an interface that can combine the individual variables and copy function into a struct of variables and a single-combined copy function then this issue will go away since it will be 2 barriers for all the copyprivates. I am guessing this is an additional implementation effort and not a fundamental issue. 2 barriers and runtime call per variable in copy-private is high but I guess we can improve this in another patch.

> * I have written only a basic test for now, covering simple scalars and arrays. My plan is to add more tests, but I wanted to get some feedback first, to avoid having to rewrite the tests too much.

Since there is code in different sub-repos (mlir dialect, mlir translation, openmpirbuilder) we will need tests there eventually as well (mlir/test/Dialect/OpenMP/ops.mlir, mlir/test/Target/LLVMIR/, llvm/unittests/Frontend/).

luporl committed:
Use getTypeAsString() to build copy function name

luporl committed:
Add integration test

luporl committed:
Use ArrayRef in OMPIRBuilder's createSingle()

luporl committed:
clang-format

luporl commented:
> > * There is one thing that I've noticed a bit late. The `__kmpc_copyprivate` function uses 2 barriers, which may impact the performance of COPYPRIVATE with multiple variables. Combining the updates of all variables to use a single `__kmpc_copyprivate` call is not straightforward.
> 
> Looking at the code, if the OpenMPIRBuilder provides an interface that can combine the individual variables and copy function into a struct of variables and a single-combined copy function then this issue will go away since it will be 2 barriers for all the copyprivates. I am guessing this is an additional implementation effort and not a fundamental issue. 2 barriers and runtime call per variable in copy-private is high but I guess we can improve this in another patch.
> 
It seems this is what clang does. For a simple test program with 2 copyprivates
```
#pragma omp single copyprivate(i, j)
```
LLVM IR contains
```
  %.omp.copyprivate.cpr_list = alloca [2 x ptr], align 8
  ...
  %4 = getelementptr inbounds [2 x ptr], ptr %.omp.copyprivate.cpr_list, i64 0, i64 0
  store ptr %i, ptr %4, align 8
  %5 = getelementptr inbounds [2 x ptr], ptr %.omp.copyprivate.cpr_list, i64 0, i64 1
  store ptr %j, ptr %5, align 8
  %6 = load i32, ptr %.omp.copyprivate.did_it, align 4
  call void @__kmpc_copyprivate(ptr @1, i32 %1, i64 16, ptr %.omp.copyprivate.cpr_list, ptr @.omp.copyprivate.copy_func, i32 %6)
```

I would also prefer this improvement to be done in another patch.

kiranchandramohan commented:
> * I have written only a basic test for now, covering simple scalars and arrays. My plan is to add more tests, but I wanted to get some feedback first, to avoid having to rewrite the tests too much.

I think there is some additional work to support explicit-shape/vla, assumed-shape arrays and pointers/allocatables.

I am away on holiday from today evening, back on Jan 8. Is it OK to wait till then?

luporl commented:
> > * I have written only a basic test for now, covering simple scalars and arrays. My plan is to add more tests, but I wanted to get some feedback first, to avoid having to rewrite the tests too much.
> 
> I think there is some additional work to support explicit-shape/vla, assumed-shape arrays and pointers/allocatables.
> 
> I am away on holiday from today evening, back on Jan 8. Is it OK to wait till then?

Sure, no problem. I'll be away next week too, back on January 2.
Thanks for the review and helpful comments.

luporl committed:
Add missing tests

kiranchandramohan commented:
I am assuming there is still some additional work here to support some kind of arrays.

```
subroutine test_array(a)
  integer :: a(:)

  !$omp parallel private(a)
  !$omp single
  a = 100
  !$omp end single copyprivate(a)
  !$omp end parallel
end subroutine
```

```
subroutine test_array(n)
  integer :: a(n)

  !$omp parallel private(a)
  !$omp single
  a = 100
  !$omp end single copyprivate(a)
  !$omp end parallel
end subroutine
```

luporl commented:
The problem is that vla/assumed-shape arrays become `!fir.box<!fir.array<?xTYPE>>` and `!fir.box` is not accepted by the `omp` dialect (at least not by `OpenMP_PointerLikeType`).
Modifying `OpenMP.cpp` to also pass boxes to `copyprivate` results in the following error:
`'omp.single' op operand #0 must be variadic of OpenMP-compatible variable type, but got '!fir.box<!fir.array<?xi32>>'`

Is there a way to get the address of a box, to make it become a `!fir.ref<!fir.box<!fir.array<?xTYPE>>>`?
`fir.box_addr` returns only the boxed value, without the other parameters needed to perform the copy.

kiranchandramohan commented:
> The problem is that vla/assumed-shape arrays become `!fir.box<!fir.array<?xTYPE>>` and `!fir.box` is not accepted by the `omp` dialect (at least not by `OpenMP_PointerLikeType`). Modifying `OpenMP.cpp` to also pass boxes to `copyprivate` results in the following error: `'omp.single' op operand #0 must be variadic of OpenMP-compatible variable type, but got '!fir.box<!fir.array<?xi32>>'`
> 
> Is there a way to get the address of a box, to make it become a `!fir.ref<!fir.box<!fir.array<?xTYPE>>>`? `fir.box_addr` returns only the boxed value, without the other parameters needed to perform the copy.

Is it possible to create an alloca for the descriptor and store the descriptor there? This is what is done in https://github.com/llvm/llvm-project/pull/71766/files#diff-7af5cba654c125286dfc8703dfb2764d04c1c756171e74ba59e0a1666885b1b3

luporl commented:
> Is it possible to create an alloca for the descriptor and store the descriptor there? This is what is done in https://github.com/llvm/llvm-project/pull/71766/files#diff-7af5cba654c125286dfc8703dfb2764d04c1c756171e74ba59e0a1666885b1b3

This should work, I'll try to do the same for copyprivate. Thanks for pointing it out.

luporl committed:
Add support for vla/assumed-shape arrays

luporl commented:
Rebased to fix conflicts and added support for vla/assumed-shape arrays.

kiranchandramohan commented:
@luporl Thanks for the updates. Would you be OK with splitting the patch and landing them separately (semantics, OpenMP dialect, lowering flang to dialect, OpenMP IRBuilder + translation)?

luporl commented:
> @luporl Thanks for the updates. Would you be OK with splitting the patch and landing them separately (semantics, OpenMP dialect, lowering flang to dialect, OpenMP IRBuilder + translation)?

Ok, the patch ended up getting big indeed. I'll split it the way you are suggesting.

luporl commented:
Closing this PR now, as it was splitted into separate PRs that reference this one.

harishch4 commented:
> Only one thread prints 1, all the others print 0. In this case the sync variable will indeed end up in a "private" thread stack, instead of the "shared" stack. So with this approach a global sync variable would be needed instead. Thanks for catching this.

Hi @luporl Is this issue resolved? I see it is still giving the wrong output. could you please confirm?

luporl commented:
> > Only one thread prints 1, all the others print 0. In this case the sync variable will indeed end up in a "private" thread stack, instead of the "shared" stack. So with this approach a global sync variable would be needed instead. Thanks for catching this.
> 
> Hi @luporl Is this issue resolved? I see it is still giving the wrong output. could you please confirm?

@harishch4 It should be resolved, since now we're using `__kmpc_copyprivate` instead, but I'm also getting the wrong output with a flang build from yesterday. I'll investigate this issue, thanks for reporting it.

luporl commented:
It seems to me the problem is the threadprivate variable not being lowered as such inside subroutine `foo`, but the original `i` variable is being used instead.

I was able to make `copyprivate` work with the test program by manually adding `omp.threadprivate` to `foo`'s generated FIR.

And the following test program, that doesn't use `copyprivate` nor `single`, also shows the same behavior:

```
program main
  implicit none
  integer, save :: i = 0
  !$omp threadprivate(i)

  !$omp parallel
  call foo
  !$omp critical
  print *, i
  !$omp end critical
  !$omp end parallel

contains
  subroutine foo()
    i = 1
  end subroutine
end program
```

The issue also happens if I wrap the `i = 1` inside a `parallel` directive.
I'll create a new issue for this.

luporl commented:
https://github.com/llvm/llvm-project/issues/84561
